feat: introduces conditional control plane installation

bartoszmajsak opened the pull request:
## Overview

When using KfDef such as:

```
apiVersion: kfdef.apps.kubeflow.org/v1
kind: KfDef
metadata:
  name: odh-mesh
spec:
  plugins:
    - kind: KfOssmPlugin
      spec:
        mesh:
          name: minimal
          namespace: istio-system
          mode: Managed
          certificate:
            generate: true
        auth:
          name: authorino
          namespace: auth-provider
          authorino:
            label: authorino/topic=odh
  applications:
    # your components
  repos:
    - name: manifests
      uri: https://github.com/maistra/odh-manifests/tarball/v0.0.6
  version: v0.0.6
```

the operator will install a minimal version of the control plane.


## Changes

- new feature of conditional installation of control plane based on `Mesh.Mode` value
- reworked plugin integration in the `KfApp` call chain (see docs around `Apply()` and `Delete()`)
- fixed issue of leaving KfApp in an odd state when errors on Apply occurs (raised as separated PR too https://github.com/opendatahub-io/opendatahub-operator/pull/501)
- reworks SMCP readiness check - it now checks not status, but a list of components in defined states
- simplifies CRD existence check call

bartoszmajsak committed:
chore: extracts smcp GVR to a var

and reuse it in smcp-related funcs

bartoszmajsak committed:
fix: returns err on oauth creation failure

instead of nil by mistake

bartoszmajsak committed:
feat: adds ability to enable feature based on predicates

in certain cases this information can only be determined at runtime, for example based on user-defined values of the plugin spec.

bartoszmajsak committed:
feat: introduces control plane installation

coupled with wait condition

bartoszmajsak committed:
chore: reworks smcp creation

It relies on owner reference so can be cleaned up as other resources

bartoszmajsak committed:
fix: makes feature enabled by default

bartoszmajsak committed:
fix: applies cleanups only if feature is enabled

Additionally moves OssmResourceTracker creation to .Apply func

bartoszmajsak committed:
feat: reworks plugin integration to use Apply and Delete funcs of KfApp

This way we can cleanup the resources in case deploying service mesh manifests because creating those in Generate phase leaves them hanging due to failing Delete hook

bartoszmajsak committed:
fix(reconcile): corrects KfDef status in case of any error

Additionally propagates the error through reconcile to keep trying, as it was ignored.

bartoszmajsak committed:
fix: reworks SMCP component readiness check

bartoszmajsak committed:
chore: improves KfApp func docs

bartoszmajsak committed:
chore: reworks logging in verifiers

bartoszmajsak committed:
chore: simplifies CRD existence check

bartoszmajsak committed:
chore: adds tests for feature enablement

bartoszmajsak commented on pkg/kfapp/ossm/feature/builder.go:
```diff
@@ -53,6 +54,10 @@ func (fb *featureBuilder) UsingConfig(config *rest.Config) *featureBuilder {
 			return errors.WithStack(err)
 		}
 
+		if err := apiextv1.AddToScheme(f.client.Scheme()); err != nil {
```
We need this as we now query for CRD resource

cam-garrison commented on pkg/kfapp/ossm/feature/verification.go:
```diff
@@ -2,69 +2,77 @@ package feature
-		Version:  "v1",
-		Resource: "servicemeshcontrolplanes",
-	}
+const (
+	interval = 1 * time.Second
+	duration = 5 * time.Minute
+)
```
Checking every second for 5 minutes seems like overkill to me.

bartoszmajsak replied:
What should we bump here?

cam-garrison replied:
I was thinking that checking every 1 second was excessive - maybe every 5/10 seconds?

cam-garrison commented on pkg/kfapp/ossm/templates/control-plane/control-plane-managed.tmpl:
```diff
@@ -0,0 +1,19 @@
+  tracing:
+    type: None
+  addons:
+    prometheus:
+      enabled: true
```
Curious as to why Prometheus is enabled in this SMCP - in our sample in `enabling-ossm.md` we have it set to [false](https://github.com/maistra/odh-manifests/blob/682f627a4047f9cbb036d2d9d70d17b63b9a9e60/enabling-ossm.md?plain=1#L91).

bartoszmajsak replied:
There are no requirements as of now what and if self install is needed. I thought this is helpful as we have metrics from istio collected

cam-garrison replied:
Gotcha, that makes sense to me :+1:

cam-garrison reviewed (changes requested):
Requesting changes (or explanations) for two things, otherwise LGTM!

bartoszmajsak commented on pkg/kfapp/ossm/feature/verification.go:
```diff
@@ -2,69 +2,77 @@ package feature
-		Resource: "servicemeshcontrolplanes",
-	}
+const (
+	interval = 1 * time.Second
```
```suggestion
	interval = 5 * time.Second
```

bartoszmajsak committed:
fix: sets interval for SMCP polling to 5s

cam-garrison commented:
Note from testing. If the `mesh.mode` field is not set, eg:

```
mesh:
          name: basic
          namespace: istio-system
          certificate:
            generate: true
```

The operator enters a CrashLoopBackOff state, with the error 
```
2023-08-31T20:04:22.269Z INFO KfOssmPlugin Initializing {"plugin": "KfOssmPlugin"}
panic: reflect.Set: value of type string is not assignable to type ossmplugin.InstallationMode
goroutine 813 [running]:
reflect.Value.assignTo({0x1e65ba0?, 0xc0038441c0?, 0x7f992d423f18?}, {0x22e0a70, 0xb}, 0x1e5fee0, 0x0)
/usr/lib/golang/src/reflect/value.go:3064 +0x2ac
reflect.Value.Set({0x1e5fee0?, 0xc005a70180?, 0x22d975a?}, {0x1e65ba0?, 0xc0038441c0?, 0x1e5fee0?})
/usr/lib/golang/src/reflect/value.go:2090 +0xeb
github.com/opendatahub-io/opendatahub-operator/pkg/kfconfig/ossmplugin.setDefaults({0x1f9bea0?, 0xc005a70160?})
/workspace/pkg/kfconfig/ossmplugin/types.go:89 +0x2fe
github.com/opendatahub-io/opendatahub-operator/pkg/kfconfig/ossmplugin.setDefaults({0x202bea0?, 0xc005a70160?})
/workspace/pkg/kfconfig/ossmplugin/types.go:93 +0x33b
github.com/opendatahub-io/opendatahub-operator/pkg/kfconfig/ossmplugin.(*OssmPluginSpec).SetDefaults(...)
..................
```

To get out of it, have to delete KfDef and remove finalizer manually.

bartoszmajsak committed:
fix: handles types which are derived for built-in ones

for example `type InstallationMode string` was previously failing when
converting the default value defined in a custom tag. Now there's a
check performed and if value can be convereted to a given type it will.

In all other cases it will return an error.

bartoszmajsak commented:
> The operator enters a CrashLoopBackOff state, with the error

Thanks for spotting it! It's now fixed in ca7e9d8

in YAML: `mode` -> `installationMode`

values: 

  * `pre-installed`
  * `minimal`

prometheus is gone

bartoszmajsak committed:
fix: sets default control plane installation mode rely on existing
installation

Proposed installation modes are:

- minimal
- pre-installed

cam-garrison reviewed (approved):
tested some more, lgtm
